rclcpp/include/rclcpp/any_subscription_callback.hpp

```diff
@@ -476,6 +476,21 @@ class AnySubscriptionCallback
     }
   }
 
+  std::unique_ptr<ROSMessageType, ROSMessageTypeDeleter>
+  convert_custom_type_to_ros_message_unique_ptr(const SubscribedType & msg)
+  {
+    if constexpr (rclcpp::TypeAdapter<MessageT>::is_specialized::value) {
+      auto ptr = ROSMessageTypeAllocatorTraits::allocate(ros_message_type_allocator_, 1);
+      ROSMessageTypeAllocatorTraits::construct(ros_message_type_allocator_, ptr);
+      rclcpp::TypeAdapter<MessageT>::convert_to_ros_message(msg, *ptr);
+      return std::unique_ptr<ROSMessageType, ROSMessageTypeDeleter>(ptr, ros_message_type_deleter_);
+    } else {
+      throw std::runtime_error(
+              "convert_custom_type_to_ros_message_unique_ptr "
+              "unexpectedly called without TypeAdapter");
```

Comment:
I would rather show an error at build time:

```cpp
// helper added somewhere
namespace detail
{
template<class> inline constexpr bool always_false_v = false;
}
```

and then:

```suggestion
      static_assert(
          rclcpp::detail::always_false_v<ROSMessageType>,
          "convert_custom_type_to_ros_message_unique_ptr() "
          "unexpectedly called without specialized TypeAdapter");
```